libarchive/archive_write_set_format_xar.c

```diff
@@ -95,9 +97,26 @@ archive_write_set_format_xar(struct archive *_a)
 
 #else	/* Support xar format */
 
-/*#define DEBUG_PRINT_TOC		1 */
+struct xml_writer;
+static int xml_writer_create(struct xml_writer **pctx);
+static int xml_writer_start_document(struct xml_writer *ctx);
+static int xml_writer_end_document(struct xml_writer *ctx);
+static int xml_writer_set_indent(struct xml_writer *ctx, unsigned int indent);
+static int xml_writer_start_element(struct xml_writer *ctx,
+    const char *localName);
+static int xml_writer_write_attribute(struct xml_writer *ctx, const char *key,
+    const char *value);
+static int xml_writer_write_attributef(struct xml_writer *ctx, const char *key,
+    const char *format, ...);
+static int xml_writer_write_string(struct xml_writer *ctx, const char *string);
+static int xml_writer_write_base64(struct xml_writer* ctx,
+    const char *data, size_t start, size_t len);
+static int xml_writer_end_element(struct xml_writer *ctx);
+static int xml_writer_get_final_content_and_length(struct xml_writer *ctx,
+    const char **out, size_t *size);
+static int xml_writer_destroy(struct xml_writer *ctx);
```

Comment:
If it were me, I would move the whole abstraction layer to a new header, effectively removing the forward declarations. Bonus point - the debug line should no longer be considered new code :-)

Reply:
That's fair, and is a good call.

However -- in reading through some of the issues on this project and looking at some of the design decisions, I get the feeling that the maintainers value isolated, single-file abstractions like this. Nobody else needs an XML writer today, and putting it into a header namespace tells other developers that they can take a dependency on it ... which would fail to link if XAR wasn't being built!

> _Quote from Tim Kientzle on [PR 1599](https://github.com/libarchive/libarchive/pull/1599)_
> A side effect of the above principles is that we do naturally get some duplication of common basic functionality. Personally, I'm fine with that: it has the side effect of keeping most source files standalone and easy to understand without having to dig through the entire system to untangle a lot of specialized vernacular. It means that most library functionality could be implemented outside of the library without any problems.

I'm happy to let @mmatuska or Tim overrule my conscience and say "yes, please split it into a separate file" however! :smile:

Reply:
I'm not sure how moving this abstraction to a header opposes/violates the "single-file abstractions" being mentioned?